app/models/submit.rb

```diff
@@ -10,4 +10,27 @@ class Submit < ApplicationRecord
   validates :contest_problem, presence: true
   validates :author, presence: true
   validates :status, presence: true
+
+  scope :contest, (lambda { |contest_id, contest_owner_id|
+    joins(contest_problem: :contest)
+    .joins(:author)
+    .joins('INNER JOIN "contest_participations" ON '\
+      '"contests"."id" = "contest_participations"."contest_id" AND '\
+      '"contest_participations"."user_id" = "submits"."author_id"')
```

Comment:
Nie znalazłem sposobu na zapisanie tego bez SQL-a. Taka forma `JOIN` z `AND` jest równoważna z poźniejszym dodaniu warunku `"contest_participations"."user_id" = "submits"."author_id"` w `WHERE`, tylko nie potrafię tego zrobić w ActiveRecord (generuje nieprawidłową składnię przy `contest_participations`).
Być może zwięźlej byłoby zrobić surowe zapytania SQL i obsługiwać całe wiersze (co de facto jest budowane przez ten kod).